File: 54-spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        int total = n * m;

        int firstRow = 0;
        int lastRow = n - 1;
        int firstCol = 0;
        int lastCol = m - 1;

        int count = 0;
        vector<int> ans;
        while (count < total) {
            for (int i = firstCol; count < total && i <= lastCol; i++) {
                ans.push_back(matrix[firstRow][i]);
                ++count;
            }
            firstRow++;

            for (int i = firstRow; count < total && i <= lastRow; i++) {
                ans.push_back(matrix[i][lastCol]);
                ++count;
            }
            lastCol--;

            for (int i = lastCol; count < total && i >= firstCol; i--) {
                ans.push_back(matrix[lastRow][i]);
                ++count;
            }
            lastRow--;

            for (int i = lastRow; count < total && i >= firstRow; i--) {
                ans.push_back(matrix[i][firstCol]);
                ++count;
            }
            firstCol++;
        }

        return ans;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
from typing import List

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix or not matrix[0]:
            return []
        
        m = len(matrix)
        n = len(matrix[0])
        total = m * n
        count = 0

        top, bottom = 0, m - 1
        left, right = 0, n - 1

        ans = []

        while count < total:
            for i in range(left, right + 1):
                if count < total:
                    ans.append(matrix[top][i])
                    count += 1
            top += 1

            for i in range(top, bottom + 1):
                if count < total:
                    ans.append(matrix[i][right])
                    count += 1
            right -= 1

            for i in range(right, left - 1, -1):
                if count < total:
                    ans.append(matrix[bottom][i])
                    count += 1
            bottom -= 1

            for i in range(bottom, top - 1, -1):
                if count < total:
                    ans.append(matrix[i][left])
                    count += 1
            left += 1

        return ans
